### ISSUE TITLE
[FEA] split_out related support in cudf

### ISSUE THREAD

**Issue by galipremsagar** [rapidsai/cudf#2192]
**Is your feature request related to a problem? Please describe.**
We need to have an equivalent of `pandas.util.hash_pandas_object`. This is because to use split_out feature(https://examples.dask.org/dataframes/02-groupby.html#Many-groups) in dask groupby we will face an exception since we don't have one like `hash_pandas_object` implemented on our side. This might require changes on dask side once this is implemented on cudf side.

Exception is below:
```python
/conda/envs/cudf/lib/python3.7/site-packages/six.py in reraise(tp, value, tb)
    690                 value = tp()
    691             if value.__traceback__ is not tb:
--> 692                 raise value.with_traceback(tb)
    693             raise value
    694         finally:

/conda/envs/cudf/lib/python3.7/site-packages/dask/dataframe/core.py in hash_shard()
   3550     else:
   3551         h = df
-> 3552     h = hash_pandas_object(h, index=False)
   3553     if is_series_like(h):
   3554         h = h._values

/conda/envs/cudf/lib/python3.7/site-packages/pandas/core/util/hashing.py in hash_pandas_object()
    114         h = Series(h, index=obj.index, dtype='uint64', copy=False)
    115     else:
--> 116         raise TypeError("Unexpected type for hashing %s" % type(obj))
    117     return h
    118 

TypeError: Unexpected type for hashing <class 'cudf.dataframe.index.GenericIndex'>
```
**Describe the solution you'd like**
We will need to implement equivalent of `hash_pandas_object` in cudf.

**Additional context**
This is to get `split_out` feature working for larger groupby's which OOM: https://examples.dask.org/dataframes/02-groupby.html#Many-groups

**Comment by kkraus14** [comment]
Talked to @mrocklin because I'm sitting next to him and we both believe generalizing the `hash_pandas_object` in Dask should be very doable, and exposing that API in cuDF Python should be trivial.

**Comment by galipremsagar** [comment]
> Talked to @mrocklin because I'm sitting next to him and we both believe generalizing the `hash_pandas_object` in Dask should be very doable, and exposing that API in cuDF Python should be trivial.

Sure, I can take this up if it is not yet assigned to anyone else.

**Comment by kkraus14** [comment]
@galipremsagar sounds good!

**Comment by beckernick** [comment]
With recent changes, the following can be considered a minimal reproducer with cuDF built from source and Dask 2.5.0:

```python
import cudf
import dask_cudf
import numpy as np
​
df = cudf.DataFrame()
df['id'] = [0,1,0,1,0]
df['x'] = [0,1,2,3,4]
​
ddf = dask_cudf.from_cudf(df, npartitions=3)
ddf.groupby('id').x.mean(split_out=2).persist()
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-21-41a0e5dc10be> in <module>
      8 
      9 ddf = dask_cudf.from_cudf(df, npartitions=3)
---> 10 ddf.groupby('id').x.mean(split_out=2).persist()

/opt/conda/envs/rapids/lib/python3.7/site-packages/dask/base.py in persist(self, **kwargs)
    136         dask.base.persist
    137         """
--> 138         (result,) = persist(self, traverse=False, **kwargs)
    139         return result
    140 

/opt/conda/envs/rapids/lib/python3.7/site-packages/dask/base.py in persist(*args, **kwargs)
    627         postpersists.append((rebuild, a_keys, state))
    628 
--> 629     results = schedule(dsk, keys, **kwargs)
    630     d = dict(zip(keys, results))
    631     results2 = [r({k: d[k] for k in ks}, *s) for r, ks, s in postpersists]

/opt/conda/envs/rapids/lib/python3.7/site-packages/dask/local.py in get_sync(dsk, keys, **kwargs)
    525     """
    526     kwargs.pop("num_workers", None)  # if num_workers present, remove it
--> 527     return get_async(apply_sync, 1, dsk, keys, **kwargs)
    528 
    529 

/opt/conda/envs/rapids/lib/python3.7/site-packages/dask/local.py in get_async(apply_async, num_workers, dsk, result, cache, get_id, rerun_exceptions_locally, pack_exception, raise_exception, callbacks, dumps, loads, **kwargs)
    469             # Seed initial tasks into the thread pool
    470             while state["ready"] and len(state["running"]) < num_workers:
--> 471                 fire_task()
    472 
    473             # Main loop, wait on tasks to finish, insert new ones

/opt/conda/envs/rapids/lib/python3.7/site-packages/dask/local.py in fire_task()
    464                         pack_exception,
    465                     ),
--> 466                     callback=queue.put,
    467                 )
    468 

/opt/conda/envs/rapids/lib/python3.7/site-packages/dask/local.py in apply_sync(func, args, kwds, callback)
    514 def apply_sync(func, args=(), kwds={}, callback=None):
    515     """ A naive synchronous version of apply_async """
--> 516     res = func(*args, **kwds)
    517     if callback is not None:
    518         callback(res)

/opt/conda/envs/rapids/lib/python3.7/site-packages/dask/local.py in execute_task(key, task_info, dumps, loads, get_id, pack_exception)
    225         failed = False
    226     except BaseException as e:
--> 227         result = pack_exception(e, dumps)
    228         failed = True
    229     return key, result, failed

/opt/conda/envs/rapids/lib/python3.7/site-packages/dask/local.py in execute_task(key, task_info, dumps, loads, get_id, pack_exception)
    220     try:
    221         task, data = loads(task_info)
--> 222         result = _execute_task(task, data)
    223         id = get_id()
    224         result = dumps((result, id))

/opt/conda/envs/rapids/lib/python3.7/site-packages/dask/core.py in _execute_task(arg, cache, dsk)
    117         func, args = arg[0], arg[1:]
    118         args2 = [_execute_task(a, cache) for a in args]
--> 119         return func(*args2)
    120     elif not ishashable(arg):
    121         return arg

/opt/conda/envs/rapids/lib/python3.7/site-packages/dask/dataframe/core.py in hash_shard(df, nparts, split_out_setup, split_out_setup_kwargs)
   4471         h = df
   4472 
-> 4473     h = hash_object_dispatch(h, index=False)
   4474     if is_series_like(h):
   4475         h = h.values

/opt/conda/envs/rapids/lib/python3.7/site-packages/dask/utils.py in __call__(self, arg, *args, **kwargs)
    502         Call the corresponding method based on type of argument.
    503         """
--> 504         meth = self.dispatch(type(arg))
    505         return meth(arg, *args, **kwargs)
    506 

/opt/conda/envs/rapids/lib/python3.7/site-packages/dask/utils.py in dispatch(self, cls)
    496                 lk[cls] = lk[cls2]
    497                 return lk[cls2]
--> 498         raise TypeError("No dispatch for {0}".format(cls))
    499 
    500     def __call__(self, arg, *args, **kwargs):

TypeError: No dispatch for <class 'cudf.core.index.GenericIndex'>
```

**Comment by ayushdg** [comment]
Looks like the implementation in dask/dask#5184 differs slightly from what was implemented in dask/dask#5322 and #2809. 

For eg: Dispatch for the `hash_shard` implementation is called `hash_object_dispatch` vs `hash_df` in the newer Implementation. 

In addition to updating dask to use the new `hash_df` dispatch, there are other changes made in dask/dask#5184, that should be looked into as well.

**Comment by beckernick** [comment]
We should have dispatching for index hashing now as part of #2809 , right? @ayushdg pointed out that there may just be an inconsistency in the naming conventions for dispatching.

@rjzamora , do you think this is perhaps now just a naming misalignment?

**Comment by ayushdg** [comment]
Naming definitely needs to be updated. There might be some other issues that we might run into after updating the naming logic (which might be groupby specific and how the hashing methods are called). Some cases that might possibly be an issue:
- There was some change to `.values` and `._values` made as a part of dask/dask#5184. Cudf did not have support for the values property back then and the change might have been a workaround.
- With groupby, MultiIndex becomes becomes very relevant. I don't think #2809 support hashing multi index objects. (Unless I'm wrong 😃 ). Cudf hash_df dispatch might need to be updated to handle multi index objects properly.

**Comment by rjzamora** [comment]
I honestly missed that there was already a `hash_object_dispatch` definition when I implemented the changes in [dask#5322](https://github.com/dask/dask/pull/5322).  It looks to me like we should just drop the use of `hash_df` ind dask/cudf and focus on the dispatch approach used in `dd.utils` and `dd.backends`

**Comment by rjzamora** [comment]
@ayushdg - UInless you are already on it,  I would be happy to supmit PRs to dask and cudf to get [the above snippet](https://github.com/rapidsai/cudf/issues/2192#issuecomment-539082534) working.  It seems that behavior is correct with [these `dask` changes](https://github.com/dask/dask/compare/master...rjzamora:hash-dispatch), and [theese `cudf` changes](https://github.com/rapidsai/cudf/compare/branch-0.11...rjzamora:new-hash-dispatch).

**Comment by ayushdg** [comment]
@rjzamora Feel free to submit the PRs 😄.

**Comment by rjzamora** [comment]
[Dask#5476](https://github.com/dask/dask/pull/5476) was merged,  so please do comment on [#2994](https://github.com/rapidsai/cudf/pull/2994) if there is any test coverage or funtionality that is missing in that PR

**Comment by beckernick** [comment]
This is now resolved (by the combination of various PRs). Closing.
